File: bindings/Octave/mlibkriging/tests/MLPKriging_test.m
%% Tests for MLPKriging Octave binding.
%% Mirrors bindings/Python/pylibkriging/tests/MLPKriging_test.py
%% and bindings/R/rlibkriging/tests/testthat/test-MLPKriging.R.
1;  % mark this file as a script

f1d = @(x) 1 - 0.5 * (sin(12 * x) ./ (1 + x) + 2 * cos(7 * x) .* x.^5 + 0.7);
branin = @(x1, x2) (x2 - 5.1 / (4 * pi^2) .* x1.^2 + 5.0 / pi .* x1 - 6.0).^2 ...
                   + 10 * (1 - 1.0 / (8 * pi)) .* cos(x1) + 10;

n_failed = 0;

% -----------------------------------------------------------------------
% Test 1: 1D fit / predict / accessors (gauss)
% -----------------------------------------------------------------------
try
    X = linspace(0.01, 0.99, 10)';
    y = f1d(X);

    k = MLPKriging(y, X, [16, 8], 2, "selu", "gauss", "constant", true);
    s = k.summary();
    assert(ischar(s) && length(s) > 0);

    [mean_p, stdev_p] = k.predict(X, true, false, false);
    assert(length(mean_p) == 10);
    assert(all(isfinite(mean_p)));

    x_pred = linspace(0.01, 0.99, 50)';
    [mean2, ~] = k.predict(x_pred, true, false, false);
    rmse = sqrt(mean((mean2 - f1d(x_pred)).^2));
    assert(isfinite(rmse));

    assert(strcmp(k.kernel(), "gauss"));
    assert(strcmp(k.activation(), "selu"));
    hd = k.hidden_dims();
    assert(length(hd) == 2 && hd(1) == 16 && hd(2) == 8);
    assert(k.feature_dim() == 2);
    assert(k.sigma2() > 0);
    assert(isfinite(k.logLikelihood()));
    fprintf("  Test 1 MLP 1D OK (RMSE=%.4f)\n", rmse);
catch err
    fprintf("  Test 1 FAILED: %s\n", err.message);
    n_failed = n_failed + 1;
end

% -----------------------------------------------------------------------
% Test 2: Branin 2D matern5_2 + simulate + update
% -----------------------------------------------------------------------
try
    rand("state", 77);
    n = 20;
    X = rand(n, 2);
    y = branin(X(:, 1) * 15 - 5, X(:, 2) * 15);

    k = MLPKriging(y, X, [32, 16], 3, "selu", "matern5_2", "constant", true);

    X_test = rand(10, 2);
    [mean_p, stdev_p] = k.predict(X_test, true, false, false);
    assert(length(mean_p) == 10);
    assert(all(isfinite(mean_p)));
    assert(all(isfinite(stdev_p)));

    sims = k.simulate(int32(20), int32(42), X_test);
    assert(size(sims, 1) == 10);
    assert(size(sims, 2) == 20);
    assert(all(isfinite(sims(:))));

    rand("state", 1234);
    X_new = rand(3, 2);
    y_new = branin(X_new(:, 1) * 15 - 5, X_new(:, 2) * 15);
    k.update(y_new, X_new);

    [mean2, ~] = k.predict(X_test, true, false, false);
    assert(length(mean2) == 10);
    fprintf("  Test 2 MLP Branin 2D OK\n");
catch err
    fprintf("  Test 2 FAILED: %s\n", err.message);
    n_failed = n_failed + 1;
end

% -----------------------------------------------------------------------
% Test 3: predict with deriv
% -----------------------------------------------------------------------
try
    rand("state", 88);
    n = 20;
    X = rand(n, 2);
    y = sin(3 * X(:, 1)) + cos(4 * X(:, 2)) + 0.5 * X(:, 1) .* X(:, 2);

    k = MLPKriging(y, X, [16, 8], 2, "selu", "gauss");
    X_new = rand(5, 2);
    [mean_p, stdev_p, cov_p, mean_deriv, stdev_deriv] = k.predict(X_new, true, false, true);
    assert(size(mean_deriv, 1) == 5);
    assert(size(mean_deriv, 2) == 2);
    assert(size(stdev_deriv, 1) == 5);
    assert(size(stdev_deriv, 2) == 2);
    assert(all(isfinite(mean_deriv(:))));
    assert(all(isfinite(stdev_deriv(:))));
    fprintf("  Test 3 MLP predict-deriv OK\n");
catch err
    fprintf("  Test 3 FAILED: %s\n", err.message);
    n_failed = n_failed + 1;
end

% -----------------------------------------------------------------------
% Test 4: logLikelihoodFun + gradient
% -----------------------------------------------------------------------
try
    X = linspace(0.01, 0.99, 10)';
    y = f1d(X);
    k = MLPKriging(y, X, [16, 8], 2, "selu", "gauss");

    th = k.theta();
    ll_val = k.logLikelihood();
    assert(isfinite(ll_val));

    [ll, grad] = k.logLikelihoodFun(th, true, false);
    assert(isfinite(ll));
    assert(all(isfinite(grad)));

    [ll2, ~] = k.logLikelihoodFun(th, false, false);
    assert(isfinite(ll2));
    fprintf("  Test 4 MLP logLikelihoodFun OK\n");
catch err
    fprintf("  Test 4 FAILED: %s\n", err.message);
    n_failed = n_failed + 1;
end

% -----------------------------------------------------------------------
% Test 5: is_fitted
% -----------------------------------------------------------------------
try
    X = linspace(0.01, 0.99, 8)';
    y = f1d(X);
    k = MLPKriging(y, X, [16, 8], 2, "selu", "gauss");
    assert(k.is_fitted());
    fprintf("  Test 5 MLP is_fitted OK\n");
catch err
    fprintf("  Test 5 FAILED: %s\n", err.message);
    n_failed = n_failed + 1;
end

% -----------------------------------------------------------------------
% Test 6: Getters (X, y)
% -----------------------------------------------------------------------
try
    X = linspace(0.01, 0.99, 8)';
    y = f1d(X);
    k = MLPKriging(y, X, [16, 8], 2, "selu", "gauss");
    Xr = k.X();
    assert(size(Xr, 1) == 8);
    yr = k.y();
    assert(length(yr) == 8);
    assert(all(k.theta() > 0));
    assert(k.sigma2() > 0);
    fprintf("  Test 6 MLP Getters OK\n");
catch err
    fprintf("  Test 6 FAILED: %s\n", err.message);
    n_failed = n_failed + 1;
end

if n_failed > 0
    error("MLPKriging tests: %d failed", n_failed);
end
fprintf("MLPKriging tests: all 6 passed\n");

File: bindings/Python/pylibkriging/tests/MLPKriging_test.py
"""Tests for MLPKriging Python binding.

Mirrors bindings/R/rlibkriging/tests/testthat/test-MLPKriging.R.
"""

import numpy as np

import pylibkriging as lk


def f1d(x):
    x = np.asarray(x).ravel()
    return 1 - 0.5 * (np.sin(12 * x) / (1 + x) + 2 * np.cos(7 * x) * x ** 5 + 0.7)


def branin(x1, x2):
    a, b, cc = 1.0, 5.1 / (4 * np.pi ** 2), 5.0 / np.pi
    r, s, t = 6.0, 10.0, 1.0 / (8 * np.pi)
    return a * (x2 - b * x1 ** 2 + cc * x1 - r) ** 2 + s * (1 - t) * np.cos(x1) + s


def test_mlpkriging_1d_gauss():
    X = np.linspace(0.01, 0.99, 10).reshape(-1, 1)
    y = f1d(X)

    k = lk.MLPKriging(y, X,
                      hidden_dims=[16, 8], d_out=2,
                      activation="selu", kernel="gauss",
                      normalize=True,
                      parameters={"max_iter_adam": "300"})

    assert isinstance(k.summary(), str) and len(k.summary()) > 0

    mean, stdev, _, _, _ = k.predict(X, True, False, False)
    assert mean.shape[0] == 10
    assert np.all(np.isfinite(mean))

    x_pred = np.linspace(0.01, 0.99, 50).reshape(-1, 1)
    mean2, _, _, _, _ = k.predict(x_pred, True, False, False)
    rmse = float(np.sqrt(np.mean((mean2.ravel() - f1d(x_pred)) ** 2)))
    assert np.isfinite(rmse)

    assert k.kernel() == "gauss"
    assert k.activation() == "selu"
    assert list(k.hidden_dims()) == [16, 8]
    assert k.feature_dim() == 2
    assert k.sigma2() > 0
    assert np.isfinite(k.logLikelihood())


def test_mlpkriging_branin_matern():
    rng = np.random.default_rng(77)
    n = 20
    X = rng.uniform(0, 1, (n, 2))
    y = np.array([branin(X[i, 0] * 15 - 5, X[i, 1] * 15) for i in range(n)])

    k = lk.MLPKriging(y, X,
                      hidden_dims=[16, 8], d_out=2,
                      activation="selu", kernel="gauss",
                      normalize=True,
                      parameters={"max_iter_adam": "300"})

    X_test = rng.uniform(0, 1, (10, 2))
    mean, stdev, _, _, _ = k.predict(X_test, True, False, False)
    assert mean.shape[0] == 10
    assert np.all(np.isfinite(mean))
    assert np.all(np.isfinite(stdev))

    sims = k.simulate(20, 42, X_test)
    assert sims.shape == (10, 20)
    assert np.all(np.isfinite(sims))

    rng2 = np.random.default_rng(1234)
    X_new = rng2.uniform(0, 1, (3, 2))
    y_new = np.array([branin(X_new[i, 0] * 15 - 5, X_new[i, 1] * 15) for i in range(3)])
    k.update(y_new, X_new, refit=False)

    mean2, _, _, _, _ = k.predict(X_test, True, False, False)
    assert mean2.shape[0] == 10


def test_mlpkriging_predict_deriv():
    rng = np.random.default_rng(88)
    n = 20
    X = rng.uniform(0, 1, (n, 2))
    def f2(x1, x2):
        return np.sin(3 * x1) + np.cos(4 * x2) + 0.5 * x1 * x2
    y = np.array([f2(X[i, 0], X[i, 1]) for i in range(n)])

    k = lk.MLPKriging(y, X,
                      hidden_dims=[16, 8], d_out=2,
                      activation="selu", kernel="gauss",
                      parameters={"max_iter_adam": "100"})

    X_new = rng.uniform(0, 1, (5, 2))
    mean, stdev, cov, mean_deriv, stdev_deriv = k.predict(X_new, True, False, True)
    assert mean_deriv.shape == (5, 2)
    assert stdev_deriv.shape == (5, 2)
    assert np.all(np.isfinite(mean_deriv))
    assert np.all(np.isfinite(stdev_deriv))


def test_mlpkriging_loglikelihood_fun():
    X = np.linspace(0.01, 0.99, 10).reshape(-1, 1)
    y = f1d(X)

    k = lk.MLPKriging(y, X,
                      hidden_dims=[16, 8], d_out=2,
                      activation="selu", kernel="gauss",
                      parameters={"max_iter_adam": "200"})

    th = k.theta()
    ll_val = k.logLikelihood()
    assert np.isfinite(ll_val)

    ll, grad, _ = k.logLikelihoodFun(th, return_grad=True)
    assert np.isfinite(ll)
    assert np.all(np.isfinite(grad))

    ll2, _, _ = k.logLikelihoodFun(th, return_grad=False)
    assert np.isfinite(ll2)


def test_mlpkriging_is_fitted():
    X = np.linspace(0.01, 0.99, 8).reshape(-1, 1)
    y = f1d(X)

    k = lk.MLPKriging(y, X,
                      hidden_dims=[16, 8], d_out=2,
                      activation="selu", kernel="gauss",
                      parameters={"max_iter_adam": "100"})

    assert k.is_fitted()


def test_mlpkriging_getters():
    X = np.linspace(0.01, 0.99, 8).reshape(-1, 1)
    y = f1d(X)

    k = lk.MLPKriging(y, X,
                      hidden_dims=[16, 8], d_out=2,
                      activation="selu", kernel="gauss",
                      parameters={"max_iter_adam": "100"})

    assert k.X().shape == (8, 1)
    assert k.y().shape[0] == 8
    th = k.theta()
    assert th.size > 0 and np.all(th > 0)
    assert k.sigma2() > 0
